common/util/tree_operations.h

```diff
@@ -0,0 +1,926 @@
+// there is a member `TreeNodeTraits<T>::available` which is always true.
+//
+// Optional members:
+//
+// - `T* Parent()`:
+//   Returns pointer to a parent node or nullptr when the node is a tree root.
+//   Types with this member can be detected by checking whether
+//   `TreeNodeTraits<T>::Parent::available` is true.
+//
+// - `ValueType& Value()`:
+//   Returns reference to a value stored in the node.
+//   Types with this member can be detected by checking whether
+//   `TreeNodeTraits<T>::Parent::available` is true.
+//
+// - `subnodes_type` (typename):
+//   Container type which should be used for storing arrays of detached
+//   child nodes. It must be move-assignable to, and move-constructible from,
+//   a value returned by `Children()` method.
+//   When the type is absent, it is assumed to be the type returned by
+//   `Children()` method with reference and other modifiers removed.
+//   It is available through `TreeNodeTraits<T>::Children::container_type`.
+
+#ifndef VERIBLE_COMMON_UTIL_TREE_OPERATIONS_H_
+#define VERIBLE_COMMON_UTIL_TREE_OPERATIONS_H_
+
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <iosfwd>  // IWYU pragma: keep
+#include <iterator>
+#include <numeric>
+#include <set>
+#include <utility>
+#include <vector>
+
+#include "common/util/logging.h"
+#include "common/util/spacer.h"
+#include "common/util/type_traits.h"
+
+namespace verible {
+
+namespace tree_operations_internal {
+
+// TODO(mglb): move (Unavailable)FeatureTraits to a more generic header, or
+// replace them with something from standard library.
+
+// Type used as a placeholder for unavailable feature in type traits.
+struct UnavailableFeatureTraits {
+  static inline constexpr bool available = false;
+};
+
+// Type used as a base for available feature in type traits.
+struct FeatureTraits {
+  static inline constexpr bool available = true;
+};
+
+// TreeNodeTraits implementation details:
+
+// Alias to Node::subnodes_type if it exists.
+// Intended for use with detected_or_t.
+template <typename Node,  //
+          typename Type_ = typename Node::subnodes_type>
+using TreeNodeSubnodesType = Type_;
+
+// Defined when `Node` contains `Children()` method which returns reference to
+// a STL-like container with nodes.
+// TODO(mglb): Add support for children containers storing Node pointers.
+template <typename Node,  //
+          typename ChildrenType_ = decltype(std::declval<Node>().Children()),
+          typename =
+              std::void_t<decltype(*std::declval<Node>().Children().begin()),
+                          decltype(std::declval<Node>().Children().end())>>
+struct TreeNodeChildrenTraits : FeatureTraits {
+  // Container type which should be used for storing arrays of detached
+  // child nodes. It must be move-assignable to, and move-constructible from,
+  // a value returned by `Children()` method.
+  using container_type = verible::remove_cvref_t<
```

Comment:
Can you use https://www.cplusplus.com/reference/type_traits/remove_cv/ ?

Reply:
This code also needs to strip reference, so it would turn into a bit longer `std::remove_cv_t<std::remove_reference_t<...>>`.